plutus-contract/test/Spec/ScriptContext/V1/RequiredSigner.hs

```diff
@@ -0,0 +1,167 @@
+                             ToData (toBuiltinData), TxInfo, Validator, ValidatorHash)
+import Plutus.V1.Ledger.Contexts (txSignedBy)
+import Plutus.V1.Ledger.Scripts (ScriptError (EvaluationError), unitDatum)
+import PlutusTx qualified
+import PlutusTx.Prelude (traceIfFalse)
+import Prelude hiding (not)
+import Wallet.Emulator.Wallet (signPrivateKeys, walletToMockWallet)
+
+tests :: TestTree
+tests =
+    testGroup "required signer"
+        [ ownWallet
+        , otherWallet
+        , missingRequiredSignature
+        , otherWalletNoSigningProcess
+        , defaultBehaviour -- no need to explictly use mustBeSignedBy for own wallet
+        ]
+
+contract :: PaymentPubKey -> PaymentPubKeyHash -> Contract () Empty ContractError ()
+contract pk pkh = do
+    let lookups1 = Constraints.plutusV1TypedValidatorLookups typedValidator
+        tx1 = Constraints.mustPayToTheScript () (Ada.lovelaceValueOf 25000000)
+    ledgerTx1 <- submitTxConstraintsWith lookups1 tx1
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx1
+
+    utxos <- utxosAt scrAddress
+    let orefs = fst <$> Map.toList utxos
+        redeemer = Redeemer $ toBuiltinData $ unPaymentPubKeyHash pkh
+        lookups2 =
+            Constraints.plutusV1OtherScript validatorScript
+            <> Constraints.unspentOutputs utxos
+            <> Constraints.paymentPubKey pk
+        tx2 =
+            foldMap (\oref -> Constraints.mustSpendScriptOutput oref redeemer) orefs
+            <> Constraints.mustIncludeDatum unitDatum
+            <> Constraints.mustBeSignedBy pkh
+    logInfo @String $ "Required Signatories: " ++ show (Constraints.requiredSignatories tx2)
+    ledgerTx2 <- submitTxConstraintsWith @Void lookups2 tx2
+    awaitTxConfirmed $ Tx.getCardanoTxId ledgerTx2
+
+ownWallet :: TestTree
+ownWallet =
+    let pk  = mockWalletPaymentPubKey     w1
+        pkh = mockWalletPaymentPubKeyHash w1
+        trace = do
+            void $ Trace.activateContractWallet w1 $ contract pk pkh
+            void $ Trace.waitNSlots 3
+    in checkPredicateOptions defaultCheckOptions "own wallet's signature passes on-chain validation" (assertValidatedTransactionCount 2) (void trace)
+
+otherWallet :: TestTree
+otherWallet =
+    let pk  = mockWalletPaymentPubKey     w2
+        pkh = mockWalletPaymentPubKeyHash w2
+        trace = do
+            Trace.setSigningProcess w1 (Just $ signPrivateKeys [paymentPrivateKey $ fromJust $ walletToMockWallet w1, paymentPrivateKey $ fromJust $ walletToMockWallet w2])
+            void $ Trace.activateContractWallet w1 $ contract pk pkh
+            void $ Trace.waitNSlots 3
+    in checkPredicateOptions defaultCheckOptions "other wallet's signature passes on-chain validation" (assertValidatedTransactionCount 2) (void trace)
+
+missingRequiredSignature :: TestTree -- shouldn't this fail in the contract before script execution because it already knows it'll fail due to other mustBeSignedBy pubkey?
```

Comment:
What do you mean by "it'll fail due to other mustBeSignedBy pubkey"?

Reply:
The contract should be able to determine failure before executing the script because the required signer constraint has not been fulfilled. Why does it need to execute the script to determine phase-2 failure? If the script had a bug (i.e. didn't correctly check for required signer) then it'd pass phase-2 locally and proceed to submit, but then ledger would probably fail phase-1 because of the missing required signer. Why not catch this earlier?

Reply:
@sjoerdvisscher Quickly looking at `Ledger.Validation`, seems like we are using the ledger's phase 1 validation, right? Shouldn't it catch that the signature is missing?

Reply:
Yes, but we only show the first error and currently phase 2 emulator validation happens before phase 1 ledger validation.

Reply:
I guess fixing this simply entails running phase 1 validation first?

Reply:
Ok, found what causes this. If you look at the function `Wallet.Emulator.Chain.validateEm`, you'll notice that we run our custom validation rules before running the real ledger rules. Thus, in our custom validation rules, we don't have a rule which detects this phase1 error, that's why the script was run. By commented out our custom validation rules and running the tests, I'm getting:

```
      [WARNING] Slot 2: TxnValidationFail Phase1 124516ce2730493a63bf6af45e941162d3e33eba84a179d89890eea857f33013: CardanoLedgerValidationError "ApplyTxError [UtxowFailure (MissingRequiredSigners (fromList [KeyHash \"76aaef06f38cc98ed08ceb168ddb55bab2ea5df43a6847a99f086fc1\"]))]"
```

So when the real ledger rules are used, we are correctly getting a phase 1 error.

Therefore, I would propose you keep the current test case. We have a medium to high priority story in our backlog which is about completely removing our custom validation rules and using only the real ledger rules.

Reply:
Perfect. I'm excited for using only ledger validation. It may not be necessary to inverse the rules if it's going to be changed again in the coming weeks.

Reply:
True, but I'm proposing this so that you can fix the `assertFailedTransaction` right away in this PR. Up to you then :)

Reply:
Let's leave it as is, the way it's been behaving for a while, and I'll comment each test I expect to be changed. Will wait for the final solution.

Reply:
Left the following comment on two tests: "this should fail as phase-1 in the contract before script execution. We need to run phase-1 before phase-2 or only use ledger validation."